lib/gcf_sms_lib/gcf_sms_lib.c

```diff
@@ -0,0 +1,481 @@
+		return -NRF_EINVAL;
+	}
+	if (sms_buffers[sms_buffer_index].pdu_size == 0) {
+		return -NRF_EINVAL;
+	}
+
+	err = nrf_modem_at_printf("AT+CNMI=3,2,0,0");
+	if (err) {
+		LOG_ERR("nrf_modem_at_printf failed on AT+CNMI=3,2,0,0, buf: %s", buf);
+		return err;
+	}
+
+	/* Send AT+CMGS command to modem. */
+	err = nrf_modem_at_cmd(buf, len, "AT+CMGS=%d\r%s\x1a\0",
+			       sms_buffers[sms_buffer_index].pdu_size,
+			       sms_buffers[sms_buffer_index].data);
+	if (err) {
+		LOG_ERR("nrf_modem_at_cmd failed on %s, buf: %s", __func__, buf);
+		LOG_DBG("PDU size: %d", sms_buffers[sms_buffer_index].pdu_size);
+		LOG_HEXDUMP_DBG(sms_buffers[sms_buffer_index].data, 340, "pdu (hex):");
+		return err;
+	}
+
+	if (prefix_match("+CMGS:", buf)) {
+		buf[3] = 'S';
+		return 0;
+	}
+
+	return err;
+}
+
+static int at_cmd_callback_cmms(char *buf, size_t len, char *at_cmd)
+{
+	int err;
+
+	/* Send to modem without buffer. */
+	err = nrf_modem_at_printf(at_cmd);
+	if (err) {
+		if (err > 0) {
+			LOG_ERR("%s failed, error_type: %d, error_value: %d", at_cmd,
+				nrf_modem_at_err_type(err), nrf_modem_at_err(err));
+		}
+		return at_custome_cmd_respond(buf, len, "ERROR\r\n", err);
+	}
+
+	return at_custome_cmd_respond(buf, len, "OK\r\n");
+}
+
+static int at_cmd_callback_cmgw(char *buf, size_t len, char *at_cmd)
+{
+	char *s1 = NULL;
+	char *s2 = NULL;
+
+	/* Set */
+	if (!prefix_match("AT+CMGW=", at_cmd)) {
+		return at_custome_cmd_respond(buf, len, "ERROR\r\n");
+	}
+
+	s1 = strstr(at_cmd, "<CR>");
+	if (s1 != NULL) {
+		/* Set special char after length to use strtol to convert to string. */
+		s1[0] = '\r';
+		/* Move s1 to after <CR>. */
+		s1 = &s1[4];
+		s2 = strstr(at_cmd, "<ESC>");
+	}
+
+	if (s2 != NULL) {
+		/* Find first available SMS buffer. */
+		for (int i = 0; i < SMS_BUFFER_ARRAY_SIZE; i++) {
+			if (sms_buffers[i].pdu_size != 0) {
+				continue;
+			}
+			/* Store size. */
+			sms_buffers[i].pdu_size = strtol(&(at_cmd[8]), NULL, 10);
+			/* Verify that the buffer is large enough. */
+			if ((int)(s2 - s1) + 1 > SMS_BUFFER_PDU_SIZE_MAX) {
+				LOG_ERR("PDU buffer size is too small (%d > %d).",
+					(int)(s2 - s1) + 1, SMS_BUFFER_PDU_SIZE_MAX);
+			}
+			/* Store in SMS buffer. */
+			strncpy(sms_buffers[i].data, s1, (size_t)(s2 - s1));
+			sms_buffers[i].data[(int)(s2 - s1)] = '\0';
+			/* Make AT response. */
+			return at_custome_cmd_respond(buf, len,
+								  "+CMGW: %d\r\nOK\r\n", i);
+		}
+	}
+
+	LOG_ERR("Could not parse command: %s", at_cmd);
+
+	return at_custome_cmd_respond(buf, len, "ERROR\r\n");
+}
+
+static int at_cmd_callback_cmgf(char *buf, size_t len, char *at_cmd)
+{
+	/* Set */
+	if (prefix_match("AT+CMGF=0", at_cmd)) {
+		strncpy(sca_buff, &(at_cmd[8]), SCA_SIZE_MAX);
+		sca_buff[sizeof(sca_buff) - 1] = '\0';
+		return at_custome_cmd_respond(buf, len, "OK\r\n");
+	}
+
+	return at_custome_cmd_respond(buf, len, "ERROR\r\n");
+}
+
+/* Initialize the library. */
+int gcf_sms_lib_init(const struct device *unused)
```

Comment:
I would let `SYS_INIT` be the only way to initialize this library and delete the public header entirely.